src/withUASTEditor.js

```diff
@@ -0,0 +1,115 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import transformer, { DEFAULT_EXPAND_LEVEL } from './transformer';
+import PositionIndex from './PositionIndex';
+import {
+  hoverNodeById,
+  highlightNodeById,
+  makePositionIndexHook,
+  getNodePosition
+} from './helpers';
+
+function withUASTEditor(WrappedComponent) {
+  class UASTEditorWrapper extends Component {
+    constructor(props) {
+      super(props);
+
+      this.state = this.stateFromUast(props.uast);
+
+      this.onNodeHover = this.onNodeHover.bind(this);
+      this.onNodeToggle = this.onNodeToggle.bind(this);
+      this.onCursorChanged = this.onCursorChanged.bind(this);
+    }
+
+    componentWillReceiveProps(nextProps) {
+      if (nextProps.uast !== this.props.uast) {
+        this.setState(this.stateFromUast(nextProps.uast));
+      }
+    }
+
+    stateFromUast(json) {
+      this.posIndex = new PositionIndex();
+
+      return {
+        uast: transformer(
+          json,
+          DEFAULT_EXPAND_LEVEL,
+          makePositionIndexHook(this.posIndex)
+        ),
+        // control highlighted node
+        lastHighlighted: null,
+        // position of hovered node
+        pos: null
+      };
+    }
+
+    onNodeHover(id, prevId) {
+      const { uast } = this.state;
+      const pos = getNodePosition(uast[id]);
+
+      this.setState({
+        uast: hoverNodeById(uast, id, prevId),
+        pos
+      });
+    }
+
+    onNodeToggle(id) {
+      const { uast } = this.state;
+      const newUast = {
+        ...uast,
+        [id]: {
+          ...uast[id],
+          expanded: !uast[id].expanded
+        }
+      };
+      this.setState({ uast: newUast });
+    }
+
+    onCursorChanged(cursorPos) {
+      const pos = this.posIndex.get({
+        Line: cursorPos.line + 1,
+        Col: cursorPos.ch + 1
+      });
+      if (!pos) {
+        return;
+      }
+      const { uast, lastHighlighted } = this.state;
+      const newUast = highlightNodeById(uast, pos.id, lastHighlighted);
```

Comment:
I think I've found a bug here. If you click on the same word, the node gets un-highlighted.

One possible solution:

```diff
 export function highlightNodeById(uast, id, prevId) {
   const node = uast[id];
   const newUast = {
     ...uast,
     [id]: {
       ...node,
       highlighted: true
     }
   };
-  if (prevId !== null) {
+  if (prevId !== null && prevId !== id) {
     newUast[prevId] = {
       ...newUast[prevId],
       highlighted: false
     };
   }

```

Reply:
fixed